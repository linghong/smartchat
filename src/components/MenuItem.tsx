import React, { useState, useCallback, FC, useMemo } from 'react';
import Link from 'next/link';
import { useRouter } from 'next/router';
import {
  AiFillCaretDown,
  AiFillCaretUp,
  AiFillDelete,
  AiFillEdit
} from 'react-icons/ai';
import { FixedSizeList as List } from 'react-window';
import AutoSizer from 'react-virtualized-auto-sizer';

import { OptionType } from '@/src/types/common';

interface MenuItemProps {
  title: string;
  link?: string;
  itemList: OptionType[] | null;
  defaultOpen?: boolean;
  setIsSidebarOpen?: (isSidebarOpen: boolean) => void;
  onItemClick?: (id: string) => void;
  activeItemId?: string;
  onDeleteClick?: (id: string) => void;
  onEditClick?: (id: string) => void;
}

const MenuItem: FC<MenuItemProps> = ({
  title,
  link,
  itemList,
  defaultOpen = false,
  setIsSidebarOpen,
  onItemClick,
  activeItemId,
  onDeleteClick,
  onEditClick
}) => {
  const [isOpen, setIsOpen] = useState(defaultOpen);

  const router = useRouter();
  const isActive = link && router.pathname === link;

  const handleToggle = useCallback(() => setIsOpen(prev => !prev), []);

  const handleLinkClick = useCallback(
    async (e: React.MouseEvent<HTMLAnchorElement>) => {
      e.preventDefault();
      if (link) {
        await router.push(link);
        if (setIsSidebarOpen && window.innerWidth <= 480) {
          setIsSidebarOpen(false);
        }
      }
    },
    [link, router, setIsSidebarOpen]
  );

  const Row = useCallback(
    ({ index, style }: { index: number; style: React.CSSProperties }) => {
      if (!itemList) return null;
      const item = itemList[index];

      const handleClick = (id: string) => {
        if (onItemClick) {
          onItemClick(id);
        }
      };
      const isActiveLi = activeItemId === item.value;

      return (
        <div
          style={style}
          className={`flex justify-between items-center px-3 py-2 tracking-tight text-sm font-normal transition-colors duration-200 hover:bg-slate-600 hover:rounded focus:bg-indigo-100 ${isActiveLi ? 'bg-slate-400 text-indigo-200 rounded-sm' : 'text-slate-200'}`}
          onClick={() => handleClick(item.value)}
        >
          <span className="truncate" onClick={() => handleClick(item.value)}>
            {item.label}
          </span>
          <div className="flex space-x-1">
            <AiFillEdit
              className="cursor-pointer hover:text-indigo-300"
              onClick={() => onEditClick && onEditClick(item.value)}
            />
            <AiFillDelete
              className="cursor-pointer hover:text-red-500"
              onClick={() => onDeleteClick && onDeleteClick(item.value)}
            />
          </div>
        </div>
      );
    },
    [itemList, onItemClick, activeItemId, onDeleteClick, onEditClick]
  );

  const listHeight = useMemo(() => {
    const itemHeight = 40; // Assuming each item is 40px high
    const maxHeight = 360; // Maximum height of the list
    const calculatedHeight = itemList ? itemList.length * itemHeight : 0;
    return Math.min(calculatedHeight, maxHeight);
  }, [itemList]);

  return (
    <div className="mt-6 font-semibold">
      <div
        className={`flex justify-between items-center px-1 py-1 border-b text-slate-100 cursor-pointer transition-colors duration-200 hover:bg-slate-500 focus:bg-indigo-100 }`}
      >
        {link ? (
          <Link
            href={link}
            className={`flex-grow px-2 transition-colors duration-200 hover:text-indigo-300`}
            onClick={handleLinkClick}
          >
            {title}
          </Link>
        ) : (
          <span className="flex-grow text-slate-50 ">{title}</span>
        )}
        <button
          role="button"
          className="cursor-pointer transition-colors duration-200 hover:text-indigo-300"
          onClick={handleToggle}
          aria-label={isOpen ? 'Collapse' : 'Expand'}
        >
          {isOpen ? <AiFillCaretUp size={18} /> : <AiFillCaretDown size={18} />}
        </button>
      </div>
      {isOpen && (
        <div
          className="px-1 py-2 font-medium text-slate-200"
          style={{ height: listHeight }}
        >
          {itemList && (
            <AutoSizer>
              {({ width, height }) => (
                <List
                  width={width}
                  height={height}
                  itemCount={itemList.length}
                  itemSize={40}
                  className="custom-scrollbar"
                >
                  {Row}
                </List>
              )}
            </AutoSizer>
          )}
        </div>
      )}
    </div>
  );
};

export default MenuItem;
