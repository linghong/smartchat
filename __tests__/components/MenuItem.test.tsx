import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import { useRouter } from 'next/router'

import MenuItem from '@/src/components/MenuItem'

// Mock next/router
jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}))

describe('MenuItem Component', () => {
  const mockPush = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()

    const mockedUseRouter = useRouter as jest.Mock
    mockedUseRouter.mockReturnValue({
      pathname: '/current-path', // Set the current path for the router mock
      push: mockPush,
    })
  })

  test('renders MenuItem component with the title and link', () => {
    render(
      <MenuItem
        title="Test Title"
        link="/test-link"
        itemList={['Item 1', 'Item 2']}
      />,
    )

    expect(screen.getByText('Test Title')).toBeInTheDocument()
    expect(screen.getByText('Test Title')).toHaveAttribute('href', '/test-link')
  })

  test('renders MenuItem component with item list and toggles correctly', () => {
    render(<MenuItem title="Test Title" itemList={['Item 1', 'Item 2']} />)

    // Ensure the item list is not visible initially
    expect(screen.queryByText('Item 1')).toBeNull()
    expect(screen.queryByText('Item 2')).toBeNull()

    // Click to toggle the item list
    fireEvent.click(screen.getByRole('button'))

    // Ensure the item list is visible after toggling
    expect(screen.getByText('Item 1')).toBeInTheDocument()
    expect(screen.getByText('Item 2')).toBeInTheDocument()
  })

  test('applies active class when the link matches the current pathname', () => {
    render(
      <MenuItem
        title="Test Title"
        link="/current-path"
        itemList={['Item 1', 'Item 2']}
      />,
    )

    const menuItem = screen.getByText('Test Title').parentElement
    expect(menuItem).toHaveClass('bg-slate-400 text-indigo-200 rounded-sm')
  })

  test('does not apply active class when the link does not match the current pathname', () => {
    render(
      <MenuItem
        title="Test Title"
        link="/other-path"
        itemList={['Item 1', 'Item 2']}
      />,
    )

    const menuItem = screen.getByText('Test Title').parentElement
    expect(menuItem).not.toHaveClass('bg-slate-500 text-indigo-200 rounded-sm')
  })

  test('renders MenuItem component with default open state', () => {
    render(
      <MenuItem
        title="Test Title"
        itemList={['Item 1', 'Item 2']}
        defaultOpen={true}
      />,
    )

    // Ensure the item list is visible initially due to defaultOpen
    expect(screen.getByText('Item 1')).toBeInTheDocument()
    expect(screen.getByText('Item 2')).toBeInTheDocument()
  })

  test('it has hover and focus classes', () => {
    render(<MenuItem title="Test Title" itemList={['Item 1', 'Item 2']} />)
    const menuItem = screen.getByText('Test Title').parentElement
    expect(menuItem).toHaveClass('hover:bg-slate-500')
    expect(menuItem).toHaveClass('focus:bg-indigo-100')
  })

  test('MenuItem component matches snapshot', () => {
    const { asFragment } = render(
      <MenuItem
        title="Test Title"
        link="/test-link"
        itemList={['Item 1', 'Item 2']}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
