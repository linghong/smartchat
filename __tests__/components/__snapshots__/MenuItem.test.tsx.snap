// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem Component MenuItem component matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="mt-5 my-8 font-semibold"
  >
    <div
      class="flex justify-between items-center px-1 py-1 border-b hover:bg-slate-500 focus:bg-indigo-100 cursor-pointer text-slate-50"
    >
      <a
        class="flex-grow px-2 hover:underline "
        href="/test-link"
      >
        Test Title
      </a>
      <span
        class="cursor-pointer"
        role="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </span>
    </div>
  </li>
</DocumentFragment>
`;
